src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs

```diff
@@ -270,15 +270,27 @@ protected override void StartResource(ODataResource resource)
                     }
 
                     this.jsonLightResourceSerializer.InstanceAnnotationWriter.WriteInstanceAnnotations(parentNavLink.GetInstanceAnnotations(), parentNavLink.Name);
-                }
 
-                // Write the property name of an expanded navigation property to start the value.
-                this.jsonWriter.WriteName(parentNavLink.Name);
+                    if (!this.ShouldOmitNullValues())
+                    {
+                        // Write the property name of an expanded navigation property to start the value of null per preference setting.
+                        this.jsonWriter.WriteName(parentNavLink.Name);
+                    }
+                }
+                else
+                {
+                    // Write the property name of an expanded navigation property to start the value.
+                    this.jsonWriter.WriteName(parentNavLink.Name);
+                }
             }
 
             if (resource == null)
             {
-                this.jsonWriter.WriteValue((string)null);
+                if (!this.ShouldOmitNullValues())
+                {
+                    this.jsonWriter.WriteValue((string) null);

```

Comment:
Why not put line the WriteValue immediately after the WriteNames on lines 277 and remove lines 287-292?

Reply:
As we discussed, line 287-292 is still needed for case of parentNavLink == null.
Per your suggestion for optimization, added the writeValue + return after Line 27, as tradeoff between optimization and non-duplicated code.

---
In reply to: [230806136](https://github.com/OData/odata.net/pull/1106#discussion_r230806136) [](ancestors = 230806136)